# Data Mining project
## Members:
- Mateusz Idziejczak 155842
- Mateusz Stawicki 155900
- Anastasiya
- Baris
## Dataset
The dataset we are going to use is the [Titanic](https://www.kaggle.com/competitions/titanic) dataset from Kaggle. It contains data of 2224 passengers. The dataset is in a CSV format and contains TODO columns. The columns are: TODO

## Implementation
### Data loading
First we need to load the data from the CSV file. We will use the pandas library to do this.


In [748]:
import pandas as pd

data = pd.read_csv('data/train.csv')

### Data exploration
Now we will explore the data to see what it contains.

In [749]:
data.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Thayer)  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [750]:
data.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

### Data preprocessing
Before we can start building the model we need to preprocess the data. This includes:
- splitting the data into features and labels
- Removing columns that are not useful
- Handling missing values
- Encoding categorical variables
- Scaling the data
- Splitting the data into training and testing sets

#### Splitting the data into features and labels

In [751]:
X = data.drop('Survived', axis=1)
y = data['Survived']

#### Removing columns that are not useful
We will remove the `PassengerId`, `Name` and `Ticket` columns as they are not useful for the model.

In [752]:
X = X.drop(['PassengerId', 'Name', 'Ticket'], axis=1)

#### Handling missing values
We will fill the missing values in the `Age` and `Cabin` columns with the mean and mode respectively.

In [753]:
X['Age'] = X['Age'].fillna(X['Age'].mean())
X['Cabin'] = X['Cabin'].fillna(X['Cabin'].mode()[0])

#### Encoding categorical variables
We will encode the `Embarked` and `Sex` column using one-hot encoding.

In [754]:
columns_to_encode = ['Embarked', 'Sex']
try:
    X = pd.get_dummies(X, columns=columns_to_encode)
except KeyError:
    print('Columns already encoded')
# We will also encode the `Cabin` column using label encoding
from sklearn.preprocessing import LabelEncoder

label_encoder = LabelEncoder()
X['Cabin'] = label_encoder.fit_transform(X['Cabin'])
# or simply delete cabin column because there are too many missing values and unique values
# X = X.drop('Cabin', axis=1)

#### Scaling the data
We will scale the data using the L1 normalization.

In [755]:
from sklearn.preprocessing import Normalizer

scaler = Normalizer(norm='l1').fit(X)
X = pd.DataFrame(scaler.transform(X), columns=X.columns)

#### Splitting the data into training and testing sets
We will split the data into training and testing sets using the `train_test_split` method from the `sklearn` library.

In [756]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

#### Final data
The final data looks like this:

In [757]:
X.head()

Pclass       Age     SibSp  Parch      Fare     Cabin  Embarked_C  \
0  0.036474  0.267477  0.012158    0.0  0.088146  0.571429    0.000000   
1  0.005147  0.195591  0.005147    0.0  0.366904  0.416917    0.005147   
2  0.034914  0.302589  0.000000    0.0  0.092232  0.546989    0.000000   
3  0.006798  0.237933  0.006798    0.0  0.360979  0.373895    0.000000   
4  0.031562  0.368227  0.000000    0.0  0.084692  0.494477    0.000000   

   Embarked_Q  Embarked_S  Sex_female  Sex_male  
0         0.0    0.012158    0.000000  0.012158  
1         0.0    0.000000    0.005147  0.000000  
2         0.0    0.011638    0.011638  0.000000  
3         0.0    0.006798    0.006798  0.000000  
4         0.0    0.010521    0.000000  0.010521

### Selecting features
Now we will select the features that we will use to build the model. We will use the `SelectKBest` method from the `sklearn` library to select the best features.

In [758]:
from sklearn.feature_selection import SelectKBest
from sklearn.feature_selection import chi2

best_features = SelectKBest(score_func=chi2, k=5)
fit = best_features.fit(X, y)
dfscores = pd.DataFrame(fit.scores_)
dfcolumns = pd.DataFrame(X.columns)
feature_scores = pd.concat([dfcolumns, dfscores], axis=1)
feature_scores.columns = ['Feature', 'Score']
feature_scores = feature_scores.sort_values(by='Score', ascending=False)
feature_scores

Feature      Score
4         Fare  10.693433
1          Age   3.558354
9   Sex_female   1.223414
10    Sex_male   1.073307
0       Pclass   0.521455
5        Cabin   0.303734
8   Embarked_S   0.171158
3        Parch   0.098493
6   Embarked_C   0.088387
2        SibSp   0.057086
7   Embarked_Q   0.000507

### Model building
Now we will build the model. We will use the Perceptron and the Random Forest Classifier.

#### Perceptron

In [759]:
from sklearn.linear_model import Perceptron
from sklearn.metrics import accuracy_score

perceptron = Perceptron()
perceptron.fit(X_train, y_train)
y_pred = perceptron.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
accuracy

0.664804469273743

#### Random Forest Classifier

In [760]:
from sklearn.ensemble import RandomForestClassifier

random_forest = RandomForestClassifier()
random_forest.fit(X_train, y_train)
y_pred = random_forest.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
accuracy

0.7932960893854749

### Model evaluation
Now we will evaluate the models using the confusion matrix and the classification report.

In [761]:
from sklearn.metrics import confusion_matrix, classification_report

y_pred = perceptron.predict(X_test)
print('Perceptron')
print(confusion_matrix(y_test, y_pred))
print(classification_report(y_test, y_pred))

y_pred = random_forest.predict(X_test)
print('Random Forest Classifier')
print(confusion_matrix(y_test, y_pred))
print(classification_report(y_test, y_pred))

Perceptron
[[108  14]
 [ 46  11]]
              precision    recall  f1-score   support

           0       0.70      0.89      0.78       122
           1       0.44      0.19      0.27        57

    accuracy                           0.66       179
   macro avg       0.57      0.54      0.53       179
weighted avg       0.62      0.66      0.62       179

Random Forest Classifier
[[100  22]
 [ 15  42]]
              precision    recall  f1-score   support

           0       0.87      0.82      0.84       122
           1       0.66      0.74      0.69        57

    accuracy                           0.79       179
   macro avg       0.76      0.78      0.77       179
weighted avg       0.80      0.79      0.80       179



### Conclusion
The Random Forest Classifier model performed better than the Perceptron model. The Random Forest Classifier model had an accuracy of 0.82, while the Perceptron model had an accuracy of 0.77. The Random Forest Classifier model also had better precision, recall and F1-score than the Perceptron model. This shows that the Random Forest Classifier model is better at predicting whether a passenger survived the Titanic disaster or not.